tools/check.sh

```diff
@@ -1,109 +1,144 @@
 #!/bin/bash
 
+# This script checks that the LaTeX sources stick to the rules.
+
+failed=0
+
 # Ignore files where rules may be violated within macro definitions.
 texfiles=$(ls *.tex | grep -v macros.tex | grep -v layout.tex | grep -v tables.tex)
 texlibdesc="support.tex concepts.tex diagnostics.tex utilities.tex strings.tex containers.tex iterators.tex ranges.tex algorithms.tex numerics.tex time.tex locales.tex iostreams.tex regex.tex atomics.tex threads.tex"
 texlib="lib-intro.tex $texlibdesc"
 
-# Discover "Overfull \[hv]box" and "Reference ... undefined" messages from LaTeX.
-sed -n '/\.tex/{s/^.*\/\([-a-z0-9]\+\.tex\).*$/\1/;h};
-/Overfull [\\][hv]box\|LaTeX Warning..Reference/{x;p;x;p}' std.log |
-sed '/^.\+\.tex$/{N;s/\n/:/}' | grep . && exit 1
+# Filter that reformats the error message as a "workflow command",
+# for native handling by github actions.
+# Prefixes each line of input with "$*: ".
+function fail() {
+    # echo "::error file=app.js,line=10,col=15::Something went wrong"
+
+    # For some reason, the file/line/col information is not shown in the GUI.
+    # Make sure to leave it in the message proper.
+    ! sed 's/^\(.\+\.tex\):\([0-9]\+\):/::error file=source\/\1,line=\2::\1:\2:'"$*"': /; s/^\([^:][^:]\)/::error ::'"$*"': \1/' |
```

Comment:
I think the last sed command will be ill-formed if the `$*` message contains a `/` e.g. something like `| fail "misplaced begin/end"`

It's probably OK, but if we want to allow `/` in the errors we could use another character less likely to be used in the error string, e.g. `@`

     ! sed 's@^\(.\+\.tex\):\([0-9]\+\):@::error file=source/\1,line=\2::\1:\2:'"$*"': @; s@^\([^:][^:]\)@::error ::'"$*"': \1@' |

(N.B. this also allows `source/` instead of `source\/`)

Reply:
The trouble is that awk's regex machinery doesn't seem to allow references to sub-matches grouped by ().

But using @ is probably a good compromise.

Reply:
GNU awk's `gensub()` function supports back-refs.

Reply:
Great, but @burblebee already dislikes GNU-isms in sed, so better avoid them for such a central functionality.